### PR TITLE
Add FIPS building support

### PR DESCRIPTION
**Pull request by AlexSchultz-clumio**
Addresses #1002.

Description of Changes:

Based on #1028 which was based on #898.

This change adds a new variable called `enable_fips` which will install openssl and enable fips mode as a kernel paramter on boot. Additionally fips mode can be enabled while running make by setting `enable_fips=true` on the command line which will add `-fips` to the ami name and set `enable_fips` to `true` when building.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Testing Done**

Built an AMI and verified it is usable with EKS. Additionally verified fips is enabled by checking `sysctl crypto.fips_enabled` is returning 1.

### CODE REVIEW

File: eks-worker-al2-variables.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "containerd_version": "1.6.*",
     "creator": "{{env `USER`}}",
     "docker_version": "20.10.23-1.amzn2.0.1",
+    "enable_fips_mode": "false",
```
Comment (cartermckinnon) [inline eks-worker-al2-variables.json:19]: shorten to `enable_fips`?

File: scripts/enable_fips.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  # install and enable fips modules
+  sudo yum install -y dracut-fips openssl
+  sudo dracut -f
+
+  # enable fips in the boot command
+  sudo /sbin/grubby --update-kernel=ALL --args="fips=1"
+  sudo sed -i 's/^\(GRUB_CMDLINE_LINUX_DEFAULT=.*\)"$/\1 fips=1"/' /etc/default/grub
+
+  # rebuild grub
+  sudo grub2-mkconfig -o /etc/grub2.cfg
```
Comment (cartermckinnon) [inline scripts/enable_fips.sh:13]: I think the `grubby` command is the only required bit:
```suggestion
  sudo /sbin/grubby --update-kernel=ALL --args="fips=1"
```

File: scripts/enable_fips.sh
Comment (cartermckinnon) [inline scripts/enable_fips.sh:1]: can you make this kebob-case to match the newer stuff?

**Review by cartermckinnon — CHANGES_REQUESTED** [review]
Few things, but looks fine overall

File: scripts/enable_fips.sh (hunk 2 above)
Comment (cartermckinnon) [reply scripts/enable_fips.sh]: When I launch an instance with the latest 1.27 AMI, I see the other flags as expected:
```
cat /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-5.10.192-183.736.amzn2.x86_64 root=UUID=11ca5c14-da9a-4407-a25a-475704058dc1 ro console=tty0 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 nvme_core.io_timeout=4294967295 rd.emergency=poweroff rd.shell=0 LANG=en_US.UTF-8 KEYTABLE=us psi=1 clocksource=tsc tsc=reliable
```

Which base AMI are you using?

Comment (AlexSchultz-clumio) [reply scripts/enable_fips.sh]: My env is 1.24 at the moment.
```
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "build_region": "us-east-2"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "docker_version": "20.10.23-1.amzn2.0.1"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "kubernetes": "1.24.17/2023-09-14/bin/linux/x86_64"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "source_ami_id": "ami-0e6220111a9f780d6"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "source_ami_name": "amzn2-ami-minimal-hvm-2.0.20230926.0-x86_64-ebs"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "ssm_agent_version": "latest"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "created": "1696865624"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "containerd_version": "1.6.*"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "cni_plugin_version": "v1.2.0"
2023-10-09T09:46:25-06:00:     amazon-ebs: Adding tag: "Name": "amazon-eks-fips-node-1.24-v20231009"
```

Comment (AlexSchultz-clumio) [reply scripts/enable_fips.sh]: Oh it's related to the `sudo grub2-mkconfig -o /etc/grub2.cfg`. Let me try switching to just grubby as I took this from the other change.

Comment (AlexSchultz-clumio) [reply scripts/enable_fips.sh]: Ok so need to use grubby or grub-mkconfig but not both.

File: Makefile [hunk 3]
```diff
@@ -26,19 +26,26 @@ ifeq ($(call vercmp,$(kubernetes_version),gteq,1.25.0), true)
 	ami_component_description ?= (k8s: {{ user `kubernetes_version` }}, containerd: {{ user `containerd_version` }})
 endif
 
-OS=
-ifneq (,$(findstring al2023, $(PACKER_TEMPLATE_FILE)))
-	OS=-al2023
+ifneq ($(origin fips), undefined)
+	enable_fips=true
```
Comment (cartermckinnon) [inline Makefile:30]: I think we only need `enable_fips`, kind of confusing to have both

File: eks-worker-al2.json [hunk 4]
```diff
@@ -152,14 +153,26 @@
...
         "KUBERNETES_VERSION={{user `kubernetes_version`}}",
         "KERNEL_VERSION={{user `kernel_version`}}"
       ]
     },
+    {
+      "type": "shell",
+      "remote_folder": "{{ user `remote_folder`}}",
+      "script": "{{template_dir}}/scripts/enable-fips.sh",
+      "environment_vars": [
+        "ENABLE_FIPS_MODE={{user `enable_fips`}}"
```
Comment (cartermckinnon) [inline eks-worker-al2.json:167]: nit: the environment var name should match the packer var name

Comment (AlexSchultz-clumio) [reply eks-worker-al2.json:167]: I forgot to rename it.  let me fix that.

File: Makefile (hunk 3 above)
Comment (AlexSchultz-clumio) [reply Makefile]: Removed this logic.

File: doc/USER_GUIDE.md [hunk 5]
```diff
@@ -41,6 +41,7 @@ Users have the following options for specifying their own values:
 | `creator` | ```{{env `USER`}}``` |  |
 | `docker_version` | ```20.10.23-1.amzn2.0.1``` |  |
 | `encrypted` | ```false``` |  |
+| `enable_fips` | ```false``` |  |
```
Comment (cartermckinnon) [inline doc/USER_GUIDE.md:44]: Can you add a description for this template variable?

I also think we should remove the bit from the main README, we don't highlight any other variables there (like `cache_container_images`).

File: eks-worker-al2.json [hunk 6]
```diff
@@ -152,14 +153,26 @@
...
+      "environment_vars": [
+        "ENABLE_FIPS={{user `enable_fips`}}"
+      ]
+    },
+    {
+      "type": "shell",
+      "inline": ["sudo reboot"],
+      "expect_disconnect": true,
+      "pause_after": "90s"
+    },
```
Comment (cartermckinnon) [inline eks-worker-al2.json:175]: in light of: https://github.com/awslabs/amazon-eks-ami/pull/1462
```suggestion
    {
      "type": "shell",
      "remote_folder": "{{ user `remote_folder`}}",
      "inline": ["sudo reboot"],
      "expect_disconnect": true,
      "pause_after": "90s"
    },
```

**Review by cartermckinnon — APPROVED** [review]

**Review by mmerkes — APPROVED** [review]
LGTM

**Comment by dlew56** [comment]
How can customers leverage this to start provisioning FIPS 140-2 enabled AWS AMI's for EKS?

**Pull request merged**